lightning/src/util/ser.rs

```diff
@@ -381,6 +383,35 @@ impl Readable for BigSize {
 	}
 }
 
+/// The lightning protocol uses u16s for lengths in most cases. As our serialization framework
+/// primarily targets that, we must as well. However, because we may serialize objects that have
+/// more than 65K entries, we need to be able to store larger values. Thus, we define a variable
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
+struct U16Or64(pub u64);
+impl Writeable for U16Or64 {
+	#[inline]
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		if self.0 < 0xffff {
+			(self.0 as u16).write(writer)
+		} else {
+			0xffffu16.write(writer)?;
+			(self.0 - 0xfffe).write(writer)
```

Comment:
I think you may wanna comment on the algorithm here, explaining why we then subtract one byte less than the value we just wrote. Same thing for the deserialization logic below.